src/psyclone/domain/lfric/transformations/raise_psyir_2_lfric_kern_trans.py

```diff
@@ -0,0 +1,220 @@
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford STFC Daresbury Lab
+
+'''Raise generic PSyIR representing a kernel-layer routine to
+PSyclone kernel-layer-specific PSyIR which uses specialised classes.
+
+'''
+import re
+
+from psyclone.psyGen import Transformation
+from psyclone.psyir.nodes import Container, ScopingNode, FileContainer
+from psyclone.psyir.transformations import TransformationError
+from psyclone.domain.lfric.kernel.lfric_kernel_metadata import \
+    LFRicKernelMetadata
+from psyclone.domain.lfric.kernel.psyir import LFRicContainer
+
+
+def find_symbol(node, name):
+    '''Utility method to find the symbol associated with the supplied
+    name. The supplied node and all of its siblings are searched and the
+    symbol and its scoping node are returned if the symbol is found.
+
+    :param node: the starting node for the search.
+    :type: :py:class:`psyclone.psyir.nodes.node`
+    :param str name: the name of the symbol being searched.
+
+    :returns: a tuple containing the symbol with the same name as the \
+        supplied name and the scoping node for that symbol.
+    :rtype: Optional[(:py:class:`psyclone.psyir.symbols.datasymbol, \
+        :py:class:`psyclone.psyir.nodes.node`)]
+
+    '''
+    symbol = None
+    scoping_node = None
+    for test_node in node.walk(ScopingNode):
+        if name in test_node.symbol_table:
+            symbol = test_node.symbol_table.lookup(name)
+            scoping_node = test_node
+            break
+    return (symbol, scoping_node)
+
+
+class RaisePSyIR2LFRicKernTrans(Transformation):
+    '''Raise a generic PSyIR representation of a kernel-layer routine
+    to a PSyclone version with specialised domain-specific nodes and
+    symbols. This is currently limited to the specialisation of kernel
+    metadata.
+
+    >>> from psyclone.domain.lfric.transformations import \
+    ...      RaisePSyIR2LFRicKernTrans
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> CODE = ("""
+    ... MODULE example
+    ... TYPE, EXTENDS(kernel_type) :: compute_cu
+    ...   TYPE(go_arg), DIMENSION(4) :: meta_args = (/         &
+    ...     go_arg(GO_WRITE, GO_CU, GO_POINTWISE),             &
+    ...     go_arg(GO_READ, GO_CT, GO_STENCIL(000, 011, 000)), &
+    ...     go_arg(GO_READ, GO_GRID_AREA_T),                   &
+    ...     go_arg(GO_READ, GO_R_SCALAR, GO_POINTWISE)/)
+    ...   INTEGER :: ITERATES_OVER = GO_ALL_PTS
+    ...   INTEGER :: index_offset = GO_OFFSET_SW
+    ... CONTAINS
+    ...   PROCEDURE, NOPASS :: code => compute_cu_code
+    ... END TYPE compute_cu
+    ... contains
+    ...   subroutine compute_cu_code()
+    ...   end subroutine
+    ... end module""")
+    >>> fortran_reader = FortranReader()
+    >>> kernel_container = fortran_reader.psyir_from_source(CODE)
+    >>> trans = RaisePSyIR2LFRicKernTrans("compute_cu")
```

Comment:
In general we have a consistent interface to the apply method with a single node argument (apart from loop fusion). Therefore to avoid breaking this I added it to the constructor. However, I don't really mind so am happy to change it.